### PR TITLE
[Templating][Components] Unify and improve E2E testing infrastructure

### PR DESCRIPTION
**Pull request by javiercn**
* Unify the Templating and Components testing infrastructure.
* Enable test project discovery in the components E2E tests.

### CODE REVIEW

File: src/Components/test/E2ETest/Microsoft.AspNetCore.Components.E2ETests.csproj [hunk 1]
```diff
@@ -1,10 +1,10 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
+  <!-- Shared testing infrastructure for running E2E tests using selenium -->
+  <Import Project="$(SharedSourceRoot)E2ETesting\E2ETesting.props" />
```
Comment (javiercn) [inline src/Components/test/E2ETest/Microsoft.AspNetCore.Components.E2ETests.csproj]: This import and the targets import below is all that's needed to bring in Selenium E2E tests

**Review by ryanbrandenburg — APPROVED** [review]
These changes seem fine. You had talked about improving error detection and spiting tests into parts, I assume that's coming at a later date?

File: src/ProjectTemplates/test/MvcTemplateTest.cs [hunk 2]
```diff
@@ -1,32 +1,35 @@
...
         {
+            Project = projectFactory.CreateProject(output);
         }
 
+        public Project Project { get; }
+
         [Theory]
         [InlineData(null)]
-        [InlineData("F#", Skip = "https://github.com/aspnet/Templating/issues/673")]
+        //[InlineData("F#", Skip = "https://github.com/aspnet/Templating/issues/673")]
```
[the remark this answers is not in the document]
Comment (javiercn) [reply src/ProjectTemplates/test/MvcTemplateTest.cs]: It doesn't work when running from VS

**Review by dougbu — CHANGES_REQUESTED** [review]

File: .azure/pipelines/jobs/default-build.yml [hunk 3]
```diff
@@ -121,8 +121,11 @@ jobs:
       inputs:
         versionSpec: 10.x
   - ${{ if and(eq(parameters.installJdk, 'true'), eq(parameters.agentOs, 'Windows')) }}:
-    - powershell: ./eng/scripts/InstallJdk.ps1 '11.0.1'
+    - powershell: ./eng/scripts/InstallJdk.ps1 '11.0.1';Write-Host "##vso[task.setvariable variable=PATH;]${env:PATH};$env:JAVA_HOME";
       displayName: Install JDK 11
+  - ${{ if and(eq(parameters.installJdk, 'true'), eq(parameters.agentOs, 'Windows')) }}:
+    - powershell: Write-Host "##vso[task.setvariable variable=PATH;]${env:PATH};$env:JAVA_HOME";
+      displayName: Set JAVA Path
```
[the remark this answers is not in the document]
Comment (javiercn) [reply .azure/pipelines/jobs/default-build.yml]: @dougbu Can you help me out with pipelines? (I've never used them). I'm basically trying to set the PATH Correctly so that our checks for prerequisites work

File: src/ProjectTemplates/test/RazorComponentsTemplateTest.cs [hunk 4]
```diff
@@ -42,9 +47,12 @@ private void TestApplication(bool publish)
 
         private void TestBasicNavigation()
         {
+            // Give components.server enough time to load so that it can replace
+            // the prerendered content before we start making assertions.
+            Thread.Sleep(5000);
```
Comment (SteveSandersonMS) [inline src/ProjectTemplates/test/RazorComponentsTemplateTest.cs]: This seems a bit sketchy. It might be OK temporarily, but maybe we should extend the connected/disconnected APIs that @pranavkm was working on to provide a connection status value so it can be determined directly.

File: src/Shared/E2ETesting/BrowserFixture.cs [hunk 5]
```diff
@@ -52,9 +57,41 @@ public BrowserFixture()
...
+        public IMessageSink DiagnosticsMessageSink { get; }
+
+        public static bool HostSupportsBrowserAutomation => string.IsNullOrWhiteSpace(Environment.GetEnvironmentVariable("ASPNETCORE_BROWSER_AUTOMATION_DISABLED")) &&
+            (IsAppVeyor || (IsVSTS && RuntimeInformation.OSDescription.Contains("Microsoft Windows")) || OSSupportsEdge());
+
+        private static bool IsAppVeyor =>
+            Environment.GetEnvironmentVariables().Contains("APPVEYOR");
+
+        private static bool IsVSTS =>
+            Environment.GetEnvironmentVariables().Contains("TF_BUILD");
```
Comment (SteveSandersonMS) [inline src/Shared/E2ETesting/BrowserFixture.cs]: Is this stuff still applicable? I didn't think we were using AppVeyor anywhere. And is VSTS another name of Azure Pipelines, or is thing another thing we could remove?

Also, the logic around "is Windows" and `OSSupportsEdge` seems redundant, since we're running these tests against Chrome, and they should be cross-platform now, right?

File: src/Shared/E2ETesting/XUnitExtensions/XUnitTestCollectionRunnerWIthAssemblyFixture.cs [hunk 6]
```diff
@@ -1,4 +1,4 @@
-﻿// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (SteveSandersonMS) [inline src/Shared/E2ETesting/XUnitExtensions/XUnitTestCollectionRunnerWIthAssemblyFixture.cs]: Typo in filename: `XunitTestCollectionRunnerWIthAssemblyFixture` should be `XunitTestCollectionRunnerWithAssemblyFixture`.

Normally it's awkward to make case-only filename changes with Git on Windows, but since you're moving the file too, it should work OK here.

File: src/Shared/E2ETesting/BrowserFixture.cs (hunk 5 above)
Comment (javiercn) [reply src/Shared/E2ETesting/BrowserFixture.cs]: No, there are a couple of small things I need to rationalize yet, but for the most part this brings everything to a single codebase

**Review by SteveSandersonMS — APPROVED** [review]
Looks great to me! I'd be interested to know about whether this can now run cross-platform.

File: src/ProjectTemplates/test/RazorComponentsTemplateTest.cs (hunk 4 above)
Comment (javiercn) [reply src/ProjectTemplates/test/RazorComponentsTemplateTest.cs]: Well, ideally what I need is some API to know that components.server.js finished initializing so that I can do ((IJavasScriptExecutor)Browser).Execute("snippet") and determine it's done initializing.

**Comment by javiercn** [comment]
> Looks great to me! I'd be interested to know about whether this can now run cross-platform.

I'm looking to see if I can make it run on OS X (and potentially linux)

File: src/Shared/E2ETesting/BrowserFixture.cs [hunk 7]
```diff
@@ -1,24 +1,29 @@
...
-
-        public ILogs Logs { get; }
+        public BrowserFixture(IMessageSink diagnosticsMessageSink)
+        {
+            DiagnosticsMessageSink = diagnosticsMessageSink;
 
-        public ITestOutputHelper Output { get; set; }
+            if (!HostSupportsBrowserAutomation)
+            {
+                DiagnosticsMessageSink.OnMessage(new DiagnosticMessage("Host does not support browser automation."));
```
Comment (pranavkm) [inline src/Shared/E2ETesting/BrowserFixture.cs]: Is there another test that says how or what one would need to do to resolve this?

File: src/Shared/E2ETesting/BrowserFixture.cs [hunk 8]
```diff
@@ -52,9 +57,41 @@ public BrowserFixture()
...
         }
 
+        public IWebDriver Browser { get; }
+
+        public ILogs Logs { get; }
+
+        public IMessageSink DiagnosticsMessageSink { get; }
+
+        public static bool HostSupportsBrowserAutomation => string.IsNullOrWhiteSpace(Environment.GetEnvironmentVariable("ASPNETCORE_BROWSER_AUTOMATION_DISABLED")) &&
+            (IsAppVeyor || (IsVSTS && RuntimeInformation.OSDescription.Contains("Microsoft Windows")) || OSSupportsEdge());
```
Comment (pranavkm) [inline src/Shared/E2ETesting/BrowserFixture.cs]: Is this only meant to run on Windows?

File: src/Shared/E2ETesting/BrowserFixture.cs [hunk 9]
```diff
@@ -52,9 +57,41 @@ public BrowserFixture()
...
+
+        private static bool IsAppVeyor =>
+            Environment.GetEnvironmentVariables().Contains("APPVEYOR");
+
+        private static bool IsVSTS =>
+            Environment.GetEnvironmentVariables().Contains("TF_BUILD");
+
+        private static int GetWindowsVersion()
+        {
+            var osDescription = RuntimeInformation.OSDescription;
```
Comment (pranavkm) [inline src/Shared/E2ETesting/BrowserFixture.cs]: Can you use `Environment.OSVersion` here instead?

File: src/Shared/E2ETesting/E2ETesting.targets [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  <!-- Running prerequisites -->
+
+  <Target Name="EnsurePrerequisites" Condition="'$(EnforcePrerequisites)' == 'True'" BeforeTargets="EnsureNodeJSRestored">
+    <PropertyGroup>
+      <_PackageJson>$(MSBuildProjectDirectory)\package.json</_PackageJson>
+    </PropertyGroup>
+
+    <!-- JAVA -->
+    <Message Importance="High" Text="Ensuring JAVA is available" />
+    <Exec Command="java -version" />
```
Comment (pranavkm) [inline src/Shared/E2ETesting/E2ETesting.targets]: Can we capture the exit code and print a message that says what needs to be done resolve this?

**Review by pranavkm — APPROVED** [review]

File: .azure/pipelines/jobs/default-build.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (dougbu) [reply .azure/pipelines/jobs/default-build.yml]: @javiercn a couple of pointers
- the [sample for Java in pipelines](https://docs.microsoft.com/en-us/azure/devops/pipelines/languages/java?view=azure-devops) doesn't attempt to update `$env:Path` globally
  - instead that sample, wraps Gradlew, Java et cetera to execute the exact command needed or (in one case) updates `$env:Path` just before using Java
- at the moment, this repo behaves similarly
  - for example, see the https://github.com/aspnet/AspNetCore/blob/master/src/SignalR/clients/java/signalr/gradlew and https://github.com/aspnet/AspNetCore/blob/master/src/SignalR/clients/java/signalr/gradlew.bat scripts

Bottom line, I suggest doing something tactical in the scripts you need.

File: src/Shared/E2ETesting/E2ETesting.targets (hunk 10 above)
Comment (dougbu) [reply src/Shared/E2ETesting/E2ETesting.targets]: Why not `<Exec Command="${JAVA_HOME}/bin/java.exe -version"/>` ?

File: .azure/pipelines/jobs/default-build.yml [hunk 11]
```diff
@@ -123,6 +123,10 @@ jobs:
   - ${{ if and(eq(parameters.installJdk, 'true'), eq(parameters.agentOs, 'Windows')) }}:
     - powershell: ./eng/scripts/InstallJdk.ps1 '11.0.1'
       displayName: Install JDK 11
+    - powershell: Write-Host "##vso[task.prependpath]$env:JAVA_HOME\bin"
+      displayName: Prepend JAVA bin folder to the PATH.
```
[the remark this answers is not in the document]
Comment (javiercn) [reply .azure/pipelines/jobs/default-build.yml]: This is needed for https://github.com/aspnet/AspNetCore/pull/8188/files?utf8=%E2%9C%93&diff=unified#diff-4035093707e8bfd6d9ba2a44912dce70R9 to succeed when called as part of `yarn install`

File: src/Shared/E2ETesting/E2ETesting.targets (hunk 10 above)
Comment (javiercn) [reply src/Shared/E2ETesting/E2ETesting.targets]: Because this needs to run cross-platform and needs to be on the PATH for https://github.com/aspnet/AspNetCore/pull/8188/files?utf8=%E2%9C%93&diff=unified#diff-4035093707e8bfd6d9ba2a44912dce70R9 to work

File: .azure/pipelines/jobs/default-build.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (pranavkm) [reply .azure/pipelines/jobs/default-build.yml]: Is there a reason this is super problematic? It's only building on the CI (so local builds aren't affected) and goes away once the build finishes.

File: src/Shared/E2ETesting/BrowserFixture.cs (hunk 5 above)
Comment (javiercn) [reply src/Shared/E2ETesting/BrowserFixture.cs]: I cleaned this up a bit.

File: src/Shared/E2ETesting/BrowserFixture.cs (hunk 7 above)
Comment (javiercn) [reply src/Shared/E2ETesting/BrowserFixture.cs]: The build will say it. I'm gonna get rid of this in the next phase

File: src/Shared/E2ETesting/BrowserFixture.cs (hunk 8 above)
Comment (javiercn) [reply src/Shared/E2ETesting/BrowserFixture.cs]: For now yes, although I've cleaned up all this crap

File: src/Shared/E2ETesting/BrowserFixture.cs (hunk 9 above)
Comment (javiercn) [reply src/Shared/E2ETesting/BrowserFixture.cs]: Got rid of this alltogether

File: src/ProjectTemplates/test/MvcTemplateTest.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (javiercn) [reply src/ProjectTemplates/test/MvcTemplateTest.cs]: Undone

File: src/ProjectTemplates/test/RazorComponentsTemplateTest.cs (hunk 4 above)
Comment (javiercn) [reply src/ProjectTemplates/test/RazorComponentsTemplateTest.cs]: That's a crazy amount of work for these tests, and wouldn't work for Blazor in the future. We should let people know when we are done initializing stuff by for example, simply saving the boot promise inside the window[Blazor] object.

**Comment by javiercn** [comment]
@dougbu This is blocked on you. 

The requirements have been explained. Can you please withdrawn your request for changes? We want to get this in ASAP as we haven't been running these tests regularly in a long time and has caused us issues.

**Review by dougbu — APPROVED** [review]

**Comment by dougbu** [comment]
Note
> should clean up the Gradlew script and anything else that updates `$env:Path` for Java on Windows

isn't urgent and doesn't need to be included in this PR. If you don't include it, please file an issue.

**Comment by ryanbrandenburg** [comment]
Any chance we could get this in today? I'd like to see if it fixes any of the the flakiness for SPA template tests.

**Comment by SteveSandersonMS** [comment]
Javier mentioned he preferred to merge this on Monday, so he could be around to watch for any possible flakiness.

**Comment by javiercn** [comment]
@aspnet/brt 
image

I've seen this failing several times.

**Comment by ryanbrandenburg** [comment]
https://github.com/aspnet/AspNetCore-Internal/issues/1913
